File: src/main/java/org/magicghostvu/actorvt/context/timer/TimerData.java
package org.magicghostvu.actorvt.context.timer;

import java.util.concurrent.Future;

public sealed abstract class TimerData permits SingleTimerData, PeriodicTimerData {

    private final Object key;
    final Future<?> job;
    private final TimerManData<?> timerMan;
    public final int generation;

    TimerData(Object key, Future<?> job, TimerManData<?> timerMan, int generation) {
        this.key = key;
        this.job = job;
        this.timerMan = timerMan;
        this.generation = generation;
    }

    public void cancel() {
        timerMan.removeKey(key, true);
    }
}

File: src/main/kotlin/org/magicghostvu/actorvt/context/timer/TimerData.kt
package org.magicghostvu.actorvt.context.timer

import java.util.concurrent.Future

sealed class TimerData(private val key: Any, val job: Future<*>, private val timerMan: TimerManData<*>) {
    public fun cancel() {
        timerMan.removeKey(key, true)
    }
}

// khi message đến/hoặc quá hạn thì xoá
class SingleTimerData internal constructor(key: Any, job: Future<*>, timerMan: TimerManData<*>) :
    TimerData(key, job, timerMan)


// xoá khi bị đè hoặc là bị huỷ
class PeriodicTimerData internal constructor(key: Any, job: Future<*>, timerMan: TimerManData<*>) :
    TimerData(key, job, timerMan)